MCH: added time information to Track class

aferrero2707 opened the pull request:
The track time is computed as the average of the time of the associated digits.
If digits are not forwarded to the track finder, the track time is set as the start of the corresponding ROF.

alibuild commented:
Error while checking build/O2/fullCI for 8caebc99adc323659713dd73c3dfd5e1bdbb6a2e at 2022-11-25 06:27:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/O2/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/O2/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/O2/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10339/8caebc99adc323659713dd73c3dfd5e1bdbb6a2e/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 2259559e203bc0af8cd1d6020bf8aa20adbd4be4 at 2022-12-04 06:24:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10339-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10339/2259559e203bc0af8cd1d6020bf8aa20adbd4be4/build_O2_fullCI/pretty.html).

aferrero2707 committed:
[MCH] added time information to Track class

The track time is computed as the average of the time of the associated digits.
If digits are not forwarded to the track finder, the track time is set
to the middle value of the corresponding ROF.

When the digits are available, the error on the time estimate is set to +/- 6 BC,
otherwise it is set as the half-width of the corresponding ROF.

aferrero2707 committed:
[MCH] added unit test for interaction record conversion

The test checks the code converting between the time from the TF start
and an absolute interaction record

aferrero2707 committed:
[MCH] fixed clang formatting

aferrero2707 commented:
@pillot @aphecetche could you have a look? Thanks!

pillot commented on DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/TrackMCH.h:
```diff
@@ -107,6 +110,18 @@ class TrackMCH
   void setClusterRef(int firstClusterIdx, int nClusters) { mClusRef.set(firstClusterIdx, nClusters); }
 
+  /// get the interaction record associated to this track
+  InteractionRecord getIR(uint32_t refOrbit) const;
```
This might be misleading as the track cannot be associated one single IR. If you want to keep this function, maybe use getMeanIR or getAverageIR or getMostProbableIR?

aferrero2707 replied:
I propose to remove the function, and call it `getMeanIR()`

pillot commented on DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/TrackMCH.h:
```diff
@@ -136,8 +151,9 @@ class TrackMCH
   double mZAtMID = 0.;                 ///< z position on the MID side where the parameters are evaluated
   double mParamAtMID[SNParams] = {0.}; ///< 5 parameters: X (cm), SlopeX, Y (cm), SlopeY, q/pYZ ((GeV/c)^-1)
   double mCovAtMID[SCovSize] = {0.};   ///< reduced covariance matrix of track parameters, formated as above
+  Time mTimeMUS;                       ///< associated time in microseconds from the TF start
```
add a default initializer: Time mTimeMUS{};

aferrero2707 replied:
Done

pillot commented on DataFormats/Detectors/MUON/MCH/src/TrackMCH.cxx:
```diff
@@ -87,6 +87,28 @@ void TrackMCH::setCovariances(const TMatrixD& src, double (&dest)[SCovSize])
 }
 
+//__________________________________________________________________________
+InteractionRecord TrackMCH::getIR(uint32_t refOrbit) const
```
this can be simplified by using the existing InteractionRecord facilities, e.g.:
InteractionRecord TrackMCH::getIR(const InteractionRecord& startIR) const
{
  auto trackBCinTF = std::llround(mTimeMUS.getTimeStamp() / constants::lhc::LHCBunchSpacingMUS);
  return startIR + trackBCinTF;
}

aferrero2707 replied:
Ok, unless we completely remove the function

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -22,6 +22,7 @@
 #include <stdexcept>
 #include <string>
 #include <filesystem>
+#include <limits>
```
not needed anymore

aferrero2707 replied:
Removed

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
   }
 
  private:
+  std::pair<InteractionRecord, bool> trackBC2IR(double trackBCinTF, const ROFRecord& trackROF, uint32_t firstTForbit) const
```
this function is not needed anymore

aferrero2707 replied:
removed

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+  }
+
+  //_________________________________________________________________________________________________
+  void setTrackTimeFromROF(TrackMCH& track, const ROFRecord& trackROF, uint32_t firstTForbit) const
```
remove this function and use the one from the ROFRecord instead, to do not duplicate the code.

aferrero2707 replied:
Indeed, removed

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+  }
+
+  //_________________________________________________________________________________________________
+  void setTrackTime(TrackMCH& track, const ROFRecord& trackROF, uint32_t firstTForbit, const gsl::span<const Cluster> usedClusters, const gsl::span<const Digit> usedDigits) const
```
trackROF --> clusterROF

aferrero2707 replied:
Ok

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+      track.setTimeMUS(tMean, tErr);
+    } else {
+      // if no digits are found, compute the time directly from the track's ROF
+      LOG(warning) << "no digits found when computing the track mean time";
```
With the current code this should never happen. This check can be kept anyway as a sanity check, but I think it should produce a LOG(fatal).

aferrero2707 replied:
Ok. I have added MCH in the message to better specify which detector is generating the fatal error

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+    } else {
+      // if no digits are found, compute the time directly from the track's ROF
+      LOG(warning) << "no digits found when computing the track mean time";
+      setTrackTimeFromROF(track, trackROF, firstTForbit);
```
use e.g. track.setTimeMUS(clusterROF.getTimeMUS({0, firstTForbit}).first); instead.

aferrero2707 replied:
Done

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -198,10 +275,18 @@ class TrackFinderTask
+      if (mDigits && usedDigits) {
+        setTrackTime(mchTracks.back(), clusterROF, firstTForbit, usedClusters, *usedDigits);
+      } else {
+        setTrackTimeFromROF(mchTracks.back(), clusterROF, firstTForbit);
```
use e.g. mchTracks.back().setTimeMUS(clusterROF.getTimeMUS({0, firstTForbit}).first); instead.

aferrero2707 replied:
done

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -198,10 +275,18 @@ class TrackFinderTask
       }
+
+      // compute the track time
+      if (mDigits && usedDigits) {
```
it's enough to check mDigits

aferrero2707 replied:
ok

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+  //_________________________________________________________________________________________________
+  void setTrackTime(TrackMCH& track, const ROFRecord& trackROF, uint32_t firstTForbit, const gsl::span<const Cluster> usedClusters, const gsl::span<const Digit> usedDigits) const
+  {
+    float trackBCinTF = 0;
```
float --> double

aferrero2707 replied:
Done

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+    for (const auto& cluster : usedClusters.subspan(track.getFirstClusterIdx(), track.getNClusters())) {
+      for (const auto& digit : usedDigits.subspan(cluster.firstDigit, cluster.nDigits)) {
+        nDigits += 1;
+        trackBCinTF += (float(digit.getTime()) - trackBCinTF) / nDigits;
```
float --> double. Even if in the end to time is saved as a float, the computation must be done in double precision to avoid rounding issues (the TF duration in ns is of the order of the significand precision of a float).

aferrero2707 replied:
Good point!

pillot commented on Detectors/MUON/MCH/Workflow/src/TrackFinderSpec.cxx:
```diff
@@ -153,12 +155,87 @@ class TrackFinderTask
+
+    // set the track IR from the computed average digits time
+    if (nDigits > 0) {
+      float tMean = o2::constants::lhc::LHCBunchSpacingMUS * trackBCinTF;
```
One must add 1.5 BC to trackBCinTF to get the correct mean value, because the digit time t0 is set to the first BC of an ADC sample which last 4BC. So the correct digit time can be t0 + 0, + 1, + 2 or + 3, which is t0 + 1.5 in average.

aferrero2707 replied:
Good point!

aferrero2707 committed:
[MCH] added review comments

pillot commented on DataFormats/Detectors/MUON/MCH/src/TrackMCH.cxx:
```diff
@@ -17,7 +17,6 @@
 #include "DataFormatsMCH/TrackMCH.h"
 
 #include <cmath>
-#include <limits>
```
this library is not used in TrackFinderSpec, but it is used in the current code of TrackMCH.

alibuild commented:
Error while checking build/O2/fullCI for f44e822e90a2165b8191879bac3ce430ce899c55 at 2022-12-07 13:27:
```
[224/743] Linking CXX executable stage/bin/o2-analysis-fdd-qa
[225/743] Building CXX object DPG/Tasks/TOF/CMakeFiles/O2Physicsexe-analysis-pid-tof-skimscreation.dir/tofSkimsTableCreator.cxx.o
[226/743] Building CXX object PWGHF/TableProducer/CMakeFiles/O2Physicsexe-analysis-hf-candidate-creator-b0.dir/candidateCreatorB0.cxx.o
[227/743] Linking CXX executable stage/bin/o2-analysis-pid-tof-skimscreation
[228/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-pair-track-track.dir/femtoDreamPairTaskTrackTrack.cxx.o
[229/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-debug-track.dir/femtoDreamDebugTrack.cxx.o
[230/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-pair-track-track
[231/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-debug-track
[232/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-producer-reduced.dir/femtoDreamProducerReducedTask.cxx.o
[233/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-pair-track-v0.dir/femtoDreamPairTaskTrackV0.cxx.o
[234/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-producer-reduced
[235/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-pair-track-v0
[236/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-debug-v0.dir/femtoDreamDebugV0.cxx.o
[237/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-hash.dir/femtoDreamHashTask.cxx.o
[238/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-debug-v0
[239/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-hash
[240/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-producer.dir/femtoDreamProducerTask.cxx.o
[241/743] Building CXX object PWGCF/FemtoWorld/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-femto-world-pair-track-track.dir/femtoWorldPairTaskTrackTrack.cxx.o
[242/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-producer
[243/743] Linking CXX executable stage/bin/o2-analysis-cf-femto-world-pair-track-track
[244/743] Building CXX object PWGCF/FemtoWorld/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-femto-world-pair-track-phi.dir/femtoWorldPairTaskTrackPhi.cxx.o
[245/743] Building CXX object PWGCF/FemtoDream/CMakeFiles/O2Physicsexe-analysis-cf-femtodream-producer-v0.dir/femtoDreamProducerTaskV0Only.cxx.o
[246/743] Linking CXX executable stage/bin/o2-analysis-cf-femto-world-pair-track-phi
[247/743] Linking CXX executable stage/bin/o2-analysis-cf-femtodream-producer-v0
[248/743] Building CXX object PWGLF/Tasks/CMakeFiles/O2Physicsexe-analysis-lf-mc-spectra-efficiency.dir/mcspectraefficiency.cxx.o
[249/743] Building CXX object PWGCF/FemtoWorld/TableProducer/CMakeFiles/O2Physicsexe-analysis-cf-femto-world-producer.dir/femtoWorldProducerTask.cxx.o
[250/743] Building CXX object PWGCF/MultiparticleCorrelations/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-multiparticle-correlations-ab.dir/multiparticle-correlations-ab.cxx.o
[251/743] Linking CXX executable stage/bin/o2-analysis-cf-femto-world-producer
[252/743] Linking CXX executable stage/bin/o2-analysis-cf-multiparticle-correlations-ab
[253/743] Building CXX object PWGCF/MultiparticleCorrelations/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-multiparticle-correlations-ar.dir/multiparticle-correlations-ar.cxx.o
[254/743] Building CXX object PWGCF/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-dptdptcorrelations.dir/dptdptcorrelations.cxx.o
[255/743] Linking CXX executable stage/bin/o2-analysis-cf-multiparticle-correlations-ar
[256/743] Building CXX object PWGCF/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-flow-generic-framework.dir/flowGenericFramework.cxx.o
[257/743] Building CXX object PWGCF/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-dptdpt-filter-qa.dir/dptdptfilterqa.cxx.o
[258/743] Linking CXX executable stage/bin/o2-analysis-cf-dptdptcorrelations
[259/743] Linking CXX executable stage/bin/o2-analysis-cf-dptdpt-filter-qa
[260/743] Linking CXX executable stage/bin/o2-analysis-cf-flow-generic-framework
[261/743] Building CXX object PWGCF/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-correlations.dir/correlations.cxx.o
[262/743] Building CXX object PWGCF/Tasks/CMakeFiles/O2Physicsexe-analysis-cf-match-reco-gen.dir/match-reco-gen.cxx.o
[263/743] Linking CXX executable stage/bin/o2-analysis-cf-correlations
[264/743] Linking CXX executable stage/bin/o2-analysis-cf-match-reco-gen
[265/743] Building CXX object PWGLF/Tasks/CMakeFiles/O2Physicsexe-analysis-lf-nuclei-batask.dir/LFNucleiBATask.cxx.o
[266/743] Building CXX object PWGCF/Tutorial/CMakeFiles/O2Physicsexe-analysis-cf-cf-tutorial-0.dir/CFTutorialTask0.cxx.o
[267/743] Linking CXX executable stage/bin/o2-analysis-lf-nuclei-batask
[268/743] Building CXX object PWGCF/TableProducer/CMakeFiles/O2Physicsexe-analysis-cf-filter-correlations.dir/filterCorrelations.cxx.o
[269/743] Linking CXX executable stage/bin/o2-analysis-cf-cf-tutorial-0
[270/743] Linking CXX executable stage/bin/o2-analysis-cf-filter-correlations
[271/743] Building CXX object PWGCF/Tutorial/CMakeFiles/O2Physicsexe-analysis-cf-cf-tutorial-1.dir/CFTutorialTask1.cxx.o
[272/743] Building CXX object PWGLF/Tasks/CMakeFiles/O2Physicsexe-analysis-lf-spectra-tof-run2.dir/spectraTOFRun2.cxx.o
[273/743] Linking CXX executable stage/bin/o2-analysis-cf-cf-tutorial-1
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10339/f44e822e90a2165b8191879bac3ce430ce899c55/build_O2_fullCI/pretty.html).

aferrero2707 committed:
[MCH] added back limits header

alibuild commented:
Error while checking build/O2/fullCI for 133011ce3f3482aec98b541c48059471fb0c7589 at 2022-12-08 17:31:
```
[4021/4075] Building CXX object Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/CMakeFiles/O2lib-TPCInterpolationWorkflow.dir/src/TPCUnbinnedResidualReaderSpec.cxx.o
[4022/4075] Building CXX object Detectors/Vertexing/CMakeFiles/O2lib-DetectorsVertexing.dir/G__O2DetectorsVertexing.cxx.o
[4023/4075] Building CXX object Detectors/Vertexing/CMakeFiles/O2lib-DetectorsVertexing.dir/src/PVertexer.cxx.o
[4024/4075] Building CXX object Detectors/Vertexing/CMakeFiles/O2lib-DetectorsVertexing.dir/src/SVertexer.cxx.o
[4025/4075] Building CXX object Detectors/Vertexing/CMakeFiles/O2lib-DetectorsVertexing.dir/src/VertexTrackMatcher.cxx.o
[4026/4075] Building CXX object Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/CMakeFiles/O2lib-TPCInterpolationWorkflow.dir/src/TPCResidualReaderSpec.cxx.o
[4027/4075] Linking CXX shared library stage/lib64/libO2DetectorsVertexing.so
[4028/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-export-workflow.dir/src/FileProducer.cxx.o
[4029/4075] Linking CXX executable stage/tests/o2-test-ccdb-CcdbApiConfigParam
[4030/4075] Building CXX object Detectors/Vertexing/CMakeFiles/O2test-detectorsvertexing-DCAFitterN.dir/test/testDCAFitterN.cxx.o
[4031/4075] Linking CXX executable stage/tests/o2-test-detectorsvertexing-DCAFitterN
[4032/4075] Linking CXX shared library stage/lib64/libO2GlobalTrackingWorkflow.so
[4033/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-aodconverter.dir/src/EveConfiguration.cxx.o
[4034/4075] Linking CXX executable stage/bin/o2-globalfwd-assessment-workflow
[4035/4075] Linking CXX executable stage/bin/o2-primary-vertex-reader-workflow
[4036/4075] Building CXX object Steer/DigitizerWorkflow/CMakeFiles/O2exe-sim-digitizer-workflow.dir/src/SimpleDigitizerWorkflow.cxx.o
[4037/4075] Linking CXX executable stage/bin/o2-tpcits-match-workflow
[4038/4075] Linking CXX executable stage/bin/o2-cosmics-match-workflow
[4039/4075] Building CXX object Detectors/Filtering/CMakeFiles/O2exe-reco-filtering-workflow.dir/src/filtering-workflow.cxx.o
[4040/4075] Linking CXX executable stage/bin/o2-primary-vertexing-workflow
[4041/4075] Linking CXX executable stage/bin/o2-secondary-vertex-reader-workflow
[4042/4075] Building CXX object Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/CMakeFiles/O2exe-tpc-scdcalib-interpolation-workflow.dir/src/tpc-interpolation-workflow.cxx.o
[4043/4075] Linking CXX executable stage/bin/o2-tof-matcher-workflow
[4044/4075] Linking CXX executable stage/bin/o2-tfidinfo-writer-workflow
[4045/4075] Linking CXX executable stage/bin/o2-secondary-vertexing-workflow
[4046/4075] Linking CXX executable stage/bin/o2-tof-match-eventtime-workflow
[4047/4075] Linking CXX executable stage/bin/o2-tof-match-checker-workflow
[4048/4075] Linking CXX executable stage/bin/o2-globalfwd-matcher-workflow
[4049/4075] Linking CXX shared library stage/lib64/libO2TPCInterpolationWorkflow.so
[4050/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-aodconverter.dir/src/FileProducer.cxx.o
[4051/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-export-workflow.dir/src/EveConfiguration.cxx.o
[4052/4075] Linking CXX executable stage/bin/o2-tpc-static-map-creator
[4053/4075] Linking CXX executable stage/bin/o2-tpc-scdcalib-interpolation-workflow
[4054/4075] Building CXX object Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/CMakeFiles/O2exe-calibration-residual-aggregator.dir/src/tpc-residual-aggregator.cxx.o
[4055/4075] Linking CXX executable stage/bin/o2-sim-digitizer-workflow
[4056/4075] Linking CXX executable stage/bin/o2-calibration-residual-aggregator
[4057/4075] Building CXX object Detectors/Align/Workflow/CMakeFiles/O2exe-barrel-alignment-workflow.dir/src/barrel-alignment-workflow.cxx.o
[4058/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-export-workflow.dir/src/DetectorData.cxx.o
[4059/4075] Linking CXX executable stage/bin/o2-barrel-alignment-workflow
[4060/4075] Building CXX object Detectors/AOD/CMakeFiles/O2exe-aod-producer-workflow.dir/src/aod-producer-workflow.cxx.o
[4061/4075] Building CXX object Detectors/AOD/CMakeFiles/O2exe-aod-mc-producer-workflow.dir/src/aod-mc-producer-workflow.cxx.o
[4062/4075] Building CXX object Detectors/Filtering/CMakeFiles/O2exe-reco-filtering-workflow.dir/src/FilteringSpec.cxx.o
[4063/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-aodconverter.dir/src/DetectorData.cxx.o
[4064/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-export-workflow.dir/src/O2DPLDisplay.cxx.o
[4065/4075] Linking CXX executable stage/bin/o2-reco-filtering-workflow
[4066/4075] Building CXX object Detectors/AOD/CMakeFiles/O2exe-aod-mc-producer-workflow.dir/src/AODMcProducerWorkflowSpec.cxx.o
[4067/4075] Linking CXX executable stage/bin/o2-aod-mc-producer-workflow
[4068/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-aodconverter.dir/src/EveWorkflowHelper.cxx.o
[4069/4075] Building CXX object EventVisualisation/Workflow/CMakeFiles/O2exe-eve-export-workflow.dir/src/EveWorkflowHelper.cxx.o
[4070/4075] Linking CXX executable stage/bin/o2-eve-export-workflow
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10339/133011ce3f3482aec98b541c48059471fb0c7589/build_O2_fullCI/pretty.html).

aphecetche commented:
Oups. Forgot to squash the commits. Sorry about that.
